Add encrypt-secret command

wmanley opened the pull request:
Adds an encrypted secret to stbt.conf that will be available unencrypted during test-runs.

This depends on the `openssl` command being available on the machine.  I don't know how common this is on non-linux environments, but it'll do for now.

wmanley committed:
Add encrypt-secret command

Adds an encrypted secret to stbt.conf that will be available unencrypted
during test-runs.

This depends on the `openssl` command being available on the machine.  I
don't know how common this is on non-linux environments, but it'll do for
now.

drothlis commented on stbt_rig.py:
```diff
@@ -352,6 +354,12 @@ def argparser():
+    encrypt_secret_parser = subcommands.add_parser(
+        "encrypt-secret", help="Write encrypted value to config file")
+    encrypt_secret_parser.add_argument(
+        "key", help="Key that can be used to retrieve the secret")
```
`name` would be a better name than `key`, to avoid confusion with the encryption keys.

wmanley replied:
Good idea, done.

drothlis commented on stbt_rig.py:
```diff
@@ -545,6 +553,65 @@ def cmd_download(args, portal):
+    for section_end in range(section_end, section_start, -1):
+        if cfg[section_end - 1].strip():
+            break
+    cfg.insert(section_end, cfg_line)
```
We could sort the entries by name with something like `sort(cfg[start:end])`, or at least insert the new key in a position that would preserve sorting if they were already sorted.

wmanley replied:
Implemented.  I don't know how people will want to organise their `stbt.conf`s, but they can always rearrange themselves.

drothlis commented:
### ConfigParser.write

`ConfigParser.write()` from the stdlib isn't suitable because it doesn't preserve comments.

### configobj

[configobj] allows round-tripping ini files preserving comments. Its last release was in 2014 but on the master branch it has commits in 2020. It's in Ubuntu from 18.04 onwards.

[configobj]: https://pypi.org/project/configobj/

`ConfigObj.list_values=False` will round-trip values with commas without adding a space between the commas. i.e. everything is a string, it won't try to convert to a list.

`ConfigObj.write_empty_values=True` will write `key =` instead of `key = ""`. Whatever we choose, it will change existing values to the format we choose so the round-trip won't always be 100% identical.

It will also normalise whitespace — specifically it will add a space around the `=` sign.

Conclusion: configobj not suitable:

- We don't want to add dependencies.
- It can reformat other entries in the ini file that we haven't touched.

wmanley committed:
tests: Test stbt_rig.py encrypt-secret

drothlis commented:
Can these encrypted values have "%"? `stbt.get_config` does "basic interpolation" so we'd need to escape "%" => "%%".

wmanley committed:
encrypt-secret: Python 2.7 compatibility

wmanley commented:
> Can these encrypted values have "%"? `stbt.get_config` does "basic interpolation" so we'd need to escape "%" => "%%".

It's base64 so the only non alphanumeric chars are `+`, `/` and `=`.

wmanley committed:
encrypt-secret: Windows support

[Docs say](https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile):

> Whether the name can be used to open the file a second time, while the
> named temporary file is still open, varies across platforms (it can be
> so used on Unix; it cannot on Windows).
